src/v/raft/state_machine_manager.cc

```diff
@@ -239,6 +245,82 @@ ss::future<> state_machine_manager::stop() {
     co_await std::move(gate_f);
 }
 
+ss::future<> state_machine_manager::apply_initial_recovery_policy() {
+    auto snapshot = co_await read_initial_recovery_snapshot();
+    if (!snapshot) {
+        /**
+         * If snapshot is not available create one, it will be populated during
+         * initial recovery
+         */
+        vlog(_log.debug, "no initial recovery snapshot found");
+        snapshot.emplace(initial_recovery_snapshot{});
+    }
+    vlog(_log.debug, "Starting with initial recovery snapshot: {}", *snapshot);
+    for (auto& [name, entry] : _machines) {
+        auto it = snapshot->initial_recovery_next_offsets.find(name);
+        if (it != snapshot->initial_recovery_next_offsets.end()) {
+            vlog(
+              _log.debug,
+              "skipping initial recovery for {} state machine as it was "
+              "recovered from snapshot",
+              name);
+            entry->stm->set_next(std::max(it->second, entry->stm->next()));
+            continue;
+        }
+        // Stm needs initial recovery
+        const auto policy = entry->stm->get_initial_recovery_policy();
+        vlog(
+          _log.info,
+          "Applying '{}' initial recovery policy for '{}' state machine",
+          policy,
+          name);
+        if (policy == stm_initial_recovery_policy::read_everything) {
+            snapshot->initial_recovery_next_offsets.emplace(
+              name, model::offset{0});
+            continue;
+        }
+        /**
+         * Here we apply the skip to end policy.
+         *
+         * The policy leverages the fact that the _next offset is already
+         * established as all local snapshots from the other state machines were
+         * already applied.
+         *
+         * We must not use the log end offset here as the end offset may change
+         * with the truncation as it haven't yet been committed in Raft sense.
+         *
+         * The `_next` is safe as it is based on the last applied offset of the
+         * other state machines i.e. the `_next` offset is already committed.
+         *
+         * This policy comes with the trade off. The newly added state machines
+         * will not actually rewind to the physical end of the log. But will
+         * read a small part of it.
+         */
+        if (policy == stm_initial_recovery_policy::skip_to_end) {
+            vlog(
+              _log.info,
+              "Setting next offset: {} for: '{}' state machine as a part of "
+              "initial recovery policy.",
+              _next,
+              name);
+            entry->stm->set_next(_next);
+            /**
+             * Initial recovery finished, marked as done in the snapshot.
+             */
+            snapshot->initial_recovery_next_offsets.emplace(
+              name, entry->stm->next());
+        }
+    }
+
+    // clean up offsets from state machines that are not longer present in the
```

Comment:
This seems correct for stms that are deleted from the codebase, but not very intuitive behavior for stms that are removed and then re-added due to some configuration changes, is this the behavior that we want for them?